<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
    <style>
        p{ white-space: pre-line;}
    </style>
</head>
<body>
<p>
一. async 基本
  返回promise对象  可以使用then方法添加回调函数，当函数遇到await就会先返回，等到异步操作完成，再接着执行函数体后面的语句
  await后面需要是promist函数  如果是普通函数会直接执行
</p>

<script>
  async function getTime() {
    console.log("---getTime")
      await  setTimeout(()=>{
              console.log("3s后")
          },3000)
      console.log("打印")
  }//此函数调用会先直接先出现 打印  ,再出现3s后  因为await后
  async function getTimeForAwait() {
      console.log("---getTime")
      await new Promise((resolve,reject)=>{
          debugger
          setTimeout(resolve,3000)   //setTime  第一个参数接受functon  type判断了下resolve为方法 resolve为调用
      })
      console.log("打印")
  }//此函数会先打印3s后  再显示打印
  //getTime()
  // getTimeForAwait()
</script>
<p>
 二. async  then  处理  以及catch处理
  普通方法
</p>
<script>
    console.log("222222222222222222222")
    //普通异步函数
   function getSomeString() {
     return  new Promise((resolve,reject)=>{
         setTimeout(resolve,3000,"3s后拿到值")
     })
 }
    //普通error函数
   function getErrorString() {
       return   new Promise((resolve,reject)=>{
           throw new Error("错误")
       })
   }
   //普通异步以及error函数
  async  function printSome() {
       console.log("先取某个值")
      await getSomeString().then(result=>{
          console.log("取到的东西1:"+result)
      }).catch(e=>{
          conso.log(e)
      })

      await getErrorString().then(result=>{
          console.log("取到的东西2:"+result)
      }).catch(e=>{
          console.log("捕捉到异常"+e)
      })
   }
   debugger
   printSome()
  console.log("异步操作不等待")

</script>
<p>
  三. async  then  处理
  依次处理多个   先到先返回方式
</p>
<script>
    console.log("33333333333333333333333333")
    async   function getString(numb) {
        return    new Promise((resolve,reject)=>{
            debugger
            setTimeout(resolve,`${numb}000`,numb)
        })
    }
    let arry=['5','3','1']
     arry.forEach(async (item)=>{
         await   getString(item).then(result=>{
             console.log(result)
         })
     })//依次打印1  3  5
</script>
<!---
     async    all  处理
     同时处理多个 同步返回
     上次遇到返回多个undefined 被合并成一个  暂时忘记怎么回事了
 --->
<script>

    async   function getString2(numb) {
        return    new Promise((resolve,reject)=>{
            setTimeout(resolve,1000,numb)
        })
    }
//    //第一种写法
//     Promise.all([getString('1'),getString('2')]).then(result=>{
//         console.log(result)
//     })
//    Promise.all([getString(1),getString(),getString(),getString(),getString(2)]).then(result=>{
//        result.forEach((item=>{
//            console.log(item)
//        }))
//    })
</script>
<!---   待继续补充--->
</body>
</html>